src/test/on_finalize.rs

```diff
@@ -0,0 +1,150 @@
+fn on_finalize_should_not_change_opening_to_begin_review_stage_prematurely() {
+    build_test_externalities().execute_with(|| {
+        let mut opening_fixture = AddOpeningFixture::default();
+        opening_fixture.max_review_period_length = 2;
+        let add_opening_result = opening_fixture.add_opening();
+        let opening_id = add_opening_result.unwrap();
+
+        assert!(Hiring::begin_review(opening_id).is_ok());
+
+        let old_opening = <OpeningById<Test>>::get(opening_id);
+        <Module<Test> as OnFinalize<BlockNumber>>::on_finalize(1);
+        let new_opening = <OpeningById<Test>>::get(opening_id);
+
+        assert_eq!(old_opening, new_opening);
+    });
+}
+
+#[test]
+fn on_finalize_should_deactivate_opening_on_begin_review_stage() {
+    build_test_externalities().execute_with(|| {
+        let mut opening_fixture = AddOpeningFixture::default();
+        opening_fixture.max_review_period_length = 2;
+        let add_opening_result = opening_fixture.add_opening();
+        let opening_id = add_opening_result.unwrap();
+
+        assert!(Hiring::begin_review(opening_id).is_ok());
+
+        let old_opening = <OpeningById<Test>>::get(opening_id);
+        if let OpeningStage::Active { stage, .. } = old_opening.stage {
+            if let ActiveOpeningStage::ReviewPeriod { .. } = stage {
+                // expected
+            } else {
+                panic!("should be BeginReview")
+            }
+        } else {
+            panic!("should be Active")
+        }
+
+        <Module<Test> as OnFinalize<BlockNumber>>::on_finalize(3);
+
+        let new_opening = <OpeningById<Test>>::get(opening_id);
+        if let OpeningStage::Active { stage, .. } = new_opening.stage {
+            if let ActiveOpeningStage::Deactivated { .. } = stage {
+                // expected
+            } else {
+                panic!("should be Deactivated")
+            }
+        } else {
+            panic!("should be Active")
+        }
+    });
+}
+
+#[test]
+fn on_finalize_should_deactivate_application_with_review_period_expired_cause() {
+    build_test_externalities().execute_with(|| {
+        let mut opening_fixture = AddOpeningFixture::default();
+        opening_fixture.max_review_period_length = 2;
+        let add_opening_result = opening_fixture.add_opening();
+        let opening_id = add_opening_result.unwrap();
+
+        let application_fixture = AddApplicationFixture::default_for_opening(opening_id);
+        let app_application_result = application_fixture.add_application();
+        let application_id = app_application_result.unwrap().application_id_added;
+
+        assert!(Hiring::begin_review(opening_id).is_ok());
+
+        let old_application = <ApplicationById<Test>>::get(application_id);
+        if let ApplicationStage::Active = old_application.stage {
+            // expected
+        } else {
+            panic!("should be Active")
+        }
+        <Module<Test> as OnFinalize<BlockNumber>>::on_finalize(3);
```

Comment:
Should this be slightly strengthened, so that you actually call with arguments `1...N-1` first, and ensure that nothing happens in those steps, only in step `N` there is actual processing?

Reply:
Also, this is really `opening_fixture.max_review_period_length+1`, right? is more clear, rather than what may seem like arbitrary constant.